packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -2,596 +2,127 @@
+import 'finders.dart';
+import 'controller.dart';
 
-/// Signature for [CommonFinders.byPredicate].
-typedef bool WidgetPredicate(Widget widget);
-
-/// Signature for [CommonFinders.byElement].
-typedef bool ElementPredicate(Element element);
+/// Signature for callback to [testWidgets] and [benchmarkWidgets].
+typedef Future<Null> WidgetTesterCallback(WidgetTester widgetTester);
 
 /// Runs the [callback] inside the Flutter test environment.
 ///
 /// Use this function for testing custom [StatelessWidget]s and
 /// [StatefulWidget]s.
 ///
+/// The callback can be asynchronous (using `async`/`await` or
+/// using explicit [Future]s).
+///
+/// This function is uses the [test] function in the test package to
+/// register the given callback as a test. The callback, when run,
+/// will be given a new instance of [WidgetTester]. The [find] object
+/// provides convenient widget [Finder]s for use with the
+/// [WidgetTester].
+///
 /// Example:
 ///
-///     test('MyWidget', () {
-///        testWidgets((WidgetTester tester) {
-///          tester.pumpWidget(new MyWidget());
-///          tester.tap(find.text('Save'));
-///          expect(tester, hasWidget(find.text('Success')));
-///        });
+///     testWidgets('MyWidget', (WidgetTester tester) {
+///       tester.pumpWidget(new MyWidget());
+///       tester.tap(find.text('Save'));
+///       expect(tester, hasWidget(find.text('Success')));
 ///     });
-void testWidgets(void callback(WidgetTester widgetTester)) {
-  testElementTree((ElementTreeTester elementTreeTester) {
-    callback(new WidgetTester._(elementTreeTester));
-  });
+void testWidgets(String description, WidgetTesterCallback callback) {
+  TestWidgetsFlutterBinding binding = TestWidgetsFlutterBinding.ensureInitialized();
+  WidgetTester tester = new WidgetTester._(binding);
+  group('-', () {
+    setUp(binding.preTest);
+    test(description, () => binding.runTest(() => callback(tester)));
+    tearDown(binding.postTest);
+  }, timeout: const Timeout(const Duration(seconds: 5)));
 }
 
-/// Class that programmatically interacts with widgets and the test environment.
-class WidgetTester {
-  WidgetTester._(this.elementTreeTester);
+/// Runs the [callback] inside the Flutter benchmark environment.
+///
+/// Use this function for benchmarking custom [StatelessWidget]s and
+/// [StatefulWidget]s when you want to be able to use features from
+/// [TestWidgetsFlutterBinding]. The callback, when run, will be given
+/// a new instance of [WidgetTester]. The [find] object provides
+/// convenient widget [Finder]s for use with the [WidgetTester].
+///
+/// The callback can be asynchronous (using `async`/`await` or using
+/// explicit [Future]s). If it is, then [benchmarkWidgets] will return
+/// a [Future] that completes when the callback's does. Otherwise, it
+/// will return a widget that is always complete.
+///
+/// Benchmarks must not be run in checked mode. To avoid this, this
+/// function will assert if it is run in checked mode.
```

Comment:
yeah. not sure what to do here. will have to see if this is a problem people run into.